0005-kzen-two-party-ecdsa.md

```diff
@@ -0,0 +1,139 @@
+# Kzen Two Party ECDSA Review
+
+* Authors: Lloyd Fournier
+* Date: 2019-05-08
+
+Issue: [#897](https://github.com/comit-network/comit-rs/issues/897)
+
+## Context
+
+In order to do a [*Scriptless Scripts* atomic swap](https://lists.linuxfoundation.org/pipermail/lightning-dev/attachments/20180426/fe978423/attachment-0001.pdf) with ECDSA signatures we need to have an implementation of Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf) protocol.
+The point of this spike is to review the protocol and KZen's [rust implementation](https://github.com/KZen-networks/multi-party-ecdsa) and make notes about.
+
+## Overview
+
+The paper by Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf)is a way to generate a joint public key and to generate joint signatures under this public key.
+This is usually called signature *aggregation*.
+There's a video by the Lindell summarizing the paper here: https://www.youtube.com/watch?v=pwc_Ork-1aA
+
+Aggregated signatures are a nice thing to have.
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`).
+
+Since ECDSA signatures doesn't have a nice algebraic structure it is **very** difficult to do signature aggregation.
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a [Paillier group](https://en.wikipedia.org/wiki/Paillier_cryptosystem) rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
+
+Once the keys are generated, Paillier is used to sign roughly in the this way:
+
+1. P1 sends the Paillier encryption of their private key to P2 (this is actually done in the keygen phase)
+2. P2 uses Paillier homorphic operations with the encrypted private key (addition and multiplication) to build a new ciphertext which has P2's parts of the signature in it and sends it to P1.
+3. P1 then decrypts it and then pretty much has a valid ECDSA signature.
```

Comment:
Because P1 will still have to do some operations on the decrypted value to actually produce the valid signature.

Note I moved and expanded this section to the end.